crates/sui-adapter-transactional-tests/tests/upgrade/basic.exp

```diff
@@ -0,0 +1,12 @@
+processed 3 tasks
+
+init:
+A: object(103)
+
+task 1 'publish'. lines 6-10:
+created: object(105), object(106)
+mutated: object(104)
+
+task 2 'upgrade'. lines 12-17:
+Error: Transaction Effects Status: Publish Error, Non-zero Address. The modules in the package must have their self-addresses set to zero.
+Execution Error: ExecutionError: ExecutionError { inner: ExecutionErrorInner { kind: PublishErrorNonZeroAddress, source: Some("Publishing module M1 with non-zero address is not allowed"), command: Some(1) } }
```

Comment:
This happens because `publish_modules` adds a named address mapping for the published module:

https://github.com/MystenLabs/sui/blob/afad24284bb6bec19ff53311f16c79057f44423a/crates/sui-transactional-test-runner/src/test_adapter.rs#L390-L403

And this named address mapping is cloned and passed into the compiler when compiling (for publish and for upgrade):

https://github.com/MystenLabs/sui/blob/afad24284bb6bec19ff53311f16c79057f44423a/external-crates/move/testing-infra/transactional-test-runner/src/framework.rs#L561

So, the fix is to pass the name of the package that you are compiling to `compile_source_units` and override its address to be `0x0` in the cloned named address mapping.